org.eclipse.scanning.api/src/org/eclipse/scanning/api/device/DeviceRequestHandler.java

```diff
@@ -1,249 +1,249 @@
+	private DeviceRequest             bean;
+	private IPublisher<DeviceRequest> publisher;
+	private IScannableDeviceService   cservice;
+
+	public DeviceRequestHandler(IRunnableDeviceService  dservice,
+			              IScannableDeviceService cservice,
+			              DeviceRequest           bean,
+			              IPublisher<DeviceRequest> statusNotifier) {
+
+		this.dservice = dservice;
+		this.cservice = cservice;
+		this.bean     = bean;
+		this.publisher = statusNotifier;
+	}
+
+	@Override
+	public DeviceRequest getBean() {
+		return bean;
+	}
+
+	@Override
+	public IPublisher<DeviceRequest> getPublisher() {
+		return publisher;
+	}
+
+	@Override
+	public DeviceRequest process(DeviceRequest request) {
+		try {
+			if (request.getDeviceType()==DeviceType.SCANNABLE) {
+				processScannables(request, cservice);
+			} else {
+				processRunnables(request, dservice);
+			}
+			return request;
+
+		} catch (ModelValidationException ne) {
+			DeviceRequest error = new DeviceRequest();
+			error.merge(request);
+			error.setErrorMessage(ne.getMessage());
+			error.setErrorFieldNames(ne.getFieldNames());
+			return error;
+
+		} catch (Exception ne) {
+			ne.printStackTrace();
+			DeviceRequest error = new DeviceRequest();
+			error.merge(request);
+			error.setErrorMessage(ne.getMessage());
+			return error;
+		}
+	}
+
+	private static void processScannables(DeviceRequest request, IScannableDeviceService cservice) throws Exception {
+
+		if (request.getDeviceName()!=null) { // Named device required
+
+			IScannable<Object> device = cservice.getScannable(request.getDeviceName());
+			DeviceAction action = request.getDeviceAction();
+			if (action==DeviceAction.SET && request.getDeviceValue()!=null) {
+
+				// Not sure if using type is a good design idea.
+				if (request.getDeviceValue() instanceof MonitorRole) {
+					device.setMonitorRole((MonitorRole)request.getDeviceValue());
+				} else {
+					device.setPosition(request.getDeviceValue(), request.getPosition());
+					/* This thread is executing to set position, while it does that it
+					 * sends events over AMQ. These events queue up with no higher priority
+					 * than this call. Therefore if we return immediately it is possible
+					 * for this call to return before position events are sent out.
+					 * FUDGE warning
+					 */
+					Thread.sleep(10);
+					 /* End warning */
+				}
+			} else if (action==DeviceAction.ACTIVATE) {
+				device.setActivated((Boolean)request.getDeviceValue());
+			}
+
+
+			if (action!=null && action.isTerminate() && device instanceof ITerminatable) {
+				ITerminatable tdevice = (ITerminatable)device;
+				tdevice.terminate(action.to());
+			}
+
+			request.setDeviceValue(device.getPosition());
+
+			DeviceInformation<?> info = new DeviceInformation<Object>(device.getName());
+			merge(info, device);
+			request.addDeviceInformation(info);
+
+		} else {
+			final Collection<String> names = cservice.getScannableNames();
+			for (String name : names) {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS135)